### PR TITLE
Extend weaponspew to handle burst weapons and beam weapons

### PR DESCRIPTION
**Pull request by EatThePath**
As the title says. 

Should resolve #2409

### CODE REVIEW

File: code/weapon/weapons.cpp [hunk 1]
```diff
@@ -7943,23 +7943,41 @@ void weapon_spew_stats(WeaponSpewType type)
...
+				mprintf(("%.2f,%.2f,", wi.max_speed, wi.max_speed * wi.lifetime));
+
+			float damage;
+			if (wi.wi_flags[Weapon::Info_Flags::Beam])
+				damage = wi.damage * wi.b_info.beam_life * (1000.0f / i2fl(BEAM_DAMAGE_TIME));
+			else
+				damage = wi.damage;
+
+			float fire_rate;
+			//We need to count the length of a firing cycle and then divide by the number of shots in that cycle
```
Comment (Goober5000) [inline code/weapon/weapons.cpp:7965]: This comment describes the reciprocal of what you're doing.

File: code/weapon/weapons.cpp [hunk 2]
```diff
@@ -8020,31 +8038,49 @@ void weapon_spew_stats(WeaponSpewType type)
...
 
 		if (all_weapons || wi.wi_flags[Weapon::Info_Flags::Player_allowed])
 		{
 			mprintf(("%s\n", wi.name));
-			mprintf(("\tVelocity: %-11.0fRange: %.0f\n", wi.max_speed, wi.max_speed * wi.lifetime));
+			//Beam range is set inthe binfo and velocity isn't very relevant to them.
+			if (wi.wi_flags[Weapon::Info_Flags::Beam])
+				mprintf((",%.2f,", wi.b_info.range));
+			else
+				mprintf(("%.2f,%.2f,", wi.max_speed, wi.max_speed * wi.lifetime));
```
Comment (Goober5000) [inline code/weapon/weapons.cpp:8051]: This looks like you did a blind copy-paste of the code in the other section.  Look again: this isn't making a CSV, this is outputting weapon stats suitable for loadout descriptions.

Comment (EatThePath) [reply code/weapon/weapons.cpp:8051]: Guilty as charged, and the same with the other item. Adjustments coming shortly

**Comment by Goober5000** [comment]
For your average fire rate calculation:

`fire_rate = (wi.burst_shots / (wi.fire_wait + wi.burst_delay * (wi.burst_shots - 1)) + (1 / wi.fire_wait)) / 2;`

Is the second part correct, or should it be `(wi.burst_shots / wi.fire_wait)` instead?

File: code/weapon/weapons.cpp [hunk 3]
```diff
@@ -8020,31 +8038,49 @@ void weapon_spew_stats(WeaponSpewType type)
...
 
 		if (all_weapons || wi.wi_flags[Weapon::Info_Flags::Player_allowed])
 		{
 			mprintf(("%s\n", wi.name));
-			mprintf(("\tVelocity: %-11.0fRange: %.0f\n", wi.max_speed, wi.max_speed * wi.lifetime));
+			//Beam range is set in the b_info and velocity isn't very relevant to them.
+			if (wi.wi_flags[Weapon::Info_Flags::Beam])
+				mprintf(("\tVelocity: N/A Range: %.0f\n", wi.b_info.range));
+			else
+				mprintf(("\tVelocity: %-11.0fRange: %.0f\n", wi.max_speed, wi.max_speed* wi.lifetime));
```
Comment (Goober5000) [inline code/weapon/weapons.cpp:8051]: Note the `%-11.0f` marker which establishes the spacing.  The table line (for Subach) looks like this:
```
Subach HL-7
	Velocity: 450        Range: 900
```

For the N/A line you will need to align the spacing, so do this:
```
"\tVelocity: N/A        Range: %.0f\n"
```

**Comment by EatThePath** [comment]
> 
> 
> For your average fire rate calculation:
> 
> ```
> fire_rate = (wi.burst_shots / (wi.fire_wait + wi.burst_delay * (wi.burst_shots - 1)) + (1 / wi.fire_wait)) / 2;
> ```
> 
> Is the second part correct, or should it be `(wi.burst_shots / wi.fire_wait)` instead?



> 
> 
> For your average fire rate calculation:
> 
> ```
> fire_rate = (wi.burst_shots / (wi.fire_wait + wi.burst_delay * (wi.burst_shots - 1)) + (1 / wi.fire_wait)) / 2;
> ```
> 
> Is the second part correct, or should it be `(wi.burst_shots / wi.fire_wait)` instead?

with random bursts, a burst ranges between the tabled burst shots as a maximum and 1 shot as the minimum. The second part of the calculation is that minimum, so it should behave exactly as if there were no burst at all in that case, hence `1/wi.fire_wait`

File: code/weapon/weapons.cpp (hunk 3 above)
Comment (EatThePath) [reply code/weapon/weapons.cpp]: done. sample output:
```Vishnan Proximity Gun
	Velocity: 1000       Range: 1800
	DPS: 120 Hull, 120 Shield, 24 Subsystem
	Power Use: 1.2W      ROF: 4/s

Vishnan Gravity Spike
	Velocity: N/A        Range: 700
	DPS: 78 Hull, 78 Shield, 194 Subsystem
	Power Use: 2.5W      ROF: 1/s```

**Comment by Goober5000** [comment]
> with random bursts, a burst ranges between the tabled burst shots as a maximum and 1 shot as the minimum. The second part of the calculation is that minimum, so it should behave exactly as if there were no burst at all in that case

Ok.

**Pull request merged**